apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+    [1, 0, 1, 0, 1, 0],
+    [0, 0, 0, 1, 0, 0],
+    [0, 0, 0, 0, 0, 0],
+    [1, 1, 0, 0, 1, 0],
+    [0, 0, 0, 0, 0, 0],
+])
+>>> mat.to_sparse_csr()
+tensor(crow_indices=tensor([0, 1, 4, 5, 5, 8, 8]),
+       col_indices=tensor([2, 0, 2, 4, 3, 0, 1, 4]),
+       values=tensor([1, 1, 1, 1, 1, 1, 1, 1]),
+       size=(6, 6),
+       nnz=8,
+       layout=torch.sparse_csr)
+```
+
+This work is concerned with the block sparse row ([BSR][pytorch-docs-bsr]) layout, which is closely
+related to the CSR layout except that blocks of values are stored. This affords many of the same
+advantages as the CSR layout, with the additional feature that vectorized operations can be used to
+operate on these value blocks. Reusing the example above with the BSR layout the similarities are
+clear
+
+```python
+>>> mat.to_sparse_bsr((2, 2))
+tensor(crow_indices=tensor([0, 3, 4, 6]),
+       col_indices=tensor([0, 1, 2, 1, 0, 2]),
+       values=tensor([[[0, 0],
+                       [1, 0]],
+                        ...
+                      [[1, 0],
+                       [0, 0]]]),
+      size=(6, 6), nnz=6, layout=torch.sparse_bsr)
+```
+
+There are some drawbacks to a block sparse layout, namely we must fully materialize a block if any
+value within it is non-zero. Larger blocks are generally better for vectorized operations, but will
+have more memory overhead.  The example we have used here demonstrates this well. In
+[Figure-1][#fig-1] we have this matrix with non-zero values in yellow, and the 2x2 block size shown
+with the red grid. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/blocksparse_2x2_matplot.png"
+    name="fig-1"
+    alt="A 6x6 grid of colored squares, zero and non-zero values indicated by color. Overlaid is a
+    grid segregating the matrix into 9 2x2 blocks."
+/>
+
+## Matrix Multiplication with Sparse Tensors
+
+In PyTorch, dense matrix multiplication in eager mode will usually forward to a high performance
+math library (i.e. cuBLAS) that implements the [`GEMM`][cublas-docs-gemm]. Once we start to involve
+sparse tensors the situation becomes more complex. We must have a different `GEMM` like operation
+for every pattern of sparse/dense arguments we need to support. Here we will focus on two forms
+involving exactly one sparse argument, we have
+
+- `C(Dense) += A(Sparse) @ B(Dense)` or **DSD**
+- `C(Sparse) += A(Dense) @ B(Dense)` or **SDD**
+
+where `@` is used to represent the matrix multiplication operation.
+
+The DSD pattern is an operation with the same semantics as a dense matrix multiplication. If the
+sparse argument were converted to dense, or all implicit zeros were materialized the result should
+be equivalent. The DSD pattern can be used to compute linear layers with sparse weights. This is an
+function which is semantically equivalent to [`torch.nn.functional.linear`][pytorch-docs-linear]
+
+```python
+def linear(x, W, b=None):
+    tmp = (W @ x.transpose(-2, -1)).transpose(-2, -1)
```

Comment:
```suggestion
    tmp = (W @ x.mT).mT
```
Also, does linear return a transposed output?

Reply:
On a second thought, I guess this is inherited from back in the days, where the matmul tiling algorithms may not have been as optimised as they are nowadays.

Reply:
Linear will return `x@W^T`, since the layout of `W` is BSR and it must be on the RHS so we compute `(W @ x^T)^T`. For dense tensors it is trivial to permute the arguments and different combinations of transposition flags will change the tiling strategy appropriately, but for sparse tensors this is not as straight forward, so APIs will usually fix which argument is carries the sparse layout. 

We are working on a kernel which will support linear in its "native" form allowing the sparse tensor to be on the LHS and carry BSC layout  which is what an expression like `x @ W^T` would boil down to as the transposition on the BSR tensor yields BSC with the same indices and transposed blocks. 

Your suggestion is still a good one as the entire post deals with 2-d arrays this is a more compact way to write the example.

Reply:
note that `.mT` is exactly the same as `transpose(-2, -1)`. I think you were thinking of `.T`, which should just be used with 2-d tensors.